### PR TITLE
Merge WC_Logger changes

### PR DESCRIPTION
**Pull request by sirreal**
Implements #10468. Replaces PRs #12575 and #12340.

The `wc-logger` branch suffered from failing tests not related to the changes. This branch is the result of a rebase which will restore automated builds and a complete passing test suite.

TODO:
- [x] Refactor `WC_Admin_Log_Table_List::prepare_items` to reduce length and complexity.
- [X] Finalize design for DB log table, ~~alternatives~~ design below (ping @mikejolley).
Full level word, colorized.

word

### CODE REVIEW

**Comment by mikejolley** [comment]
I like the top one - colours for easy scanning, and meaningful labels 👍

**Comment by claudiosanches** [comment]
I'm not just happy with the order of the columns.
I think that works better using the source first, then the level, message and timestamps.

**Comment by sirreal** [comment]
@claudiosanches
> I think that works better using the source first, then the level, message and timestamps.

`source` is optional, it may be blank. I have reservations about making it the first/primary column.

**Comment by claudiosanches** [comment]
@sirreal still the source is very important to track any issue, maybe we can use the current file that triggered the log.

**Comment by sirreal** [comment]
@claudiosanches You're right, it's important to know where messages are coming from, but shouldn't that kind of information be part of the log message itself, like including a trace as part of an error message?

`source` is like a taxonomy, it serves to group related messages. In this sense an identifier is more appropriate than a particular file name, i.e. `foobar-extension` rather than `path/class-foobar-extension-someview.php`.

**Comment by mikejolley** [comment]
I like the current order @claudiosanches

**Review by mikejolley — COMMENTED** [review]
Styling looks good to me now. Added a few notes and thats it :)

**Comment by claudiosanches** [comment]
> Are you suggesting that we should try to provide a default source based on the log call site?

Yes, or make it required in the future.

File: includes/admin/class-wc-admin-log-table-list.php [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+			'ajax'     => false,
+		) );
+	}
+
+	/**
+	 * Get list views.
+	 *
+	 * @return array Views.
+	 */
+	public function get_views() {
```
Comment (sirreal) [inline includes/admin/class-wc-admin-log-table-list.php]: Source was originally a free text search, but now that we're using a dropdown filter for source, this should also be changed to match.

**Comment by sirreal** [comment]
@claudiosanches Regarding `source`

> Yes, or make it required in the future.

According to the [PSR-3 Logger spec](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-3-logger-interface.md#13-context), I don't think `source` can be required, so I ruled that out.

I implemented functionality to attempt to provide a filename-based `source` default as discussed acd338fb87868f507c21c5b18523b4b9e24e5cc3.

**Comment by mikejolley** [comment]
@claudiosanches you happy to merge this now?

**Comment by claudiosanches** [comment]
@mikejolley yep! :)
@sirreal thanks.

File: includes/wc-deprecated-functions.php [hunk 2]
```diff
@@ -867,3 +867,18 @@ function _wc_save_product_price( $product_id, $regular_price, $sale_price = '',
...
+/**
+ * Get a log file path.
+ *
+ * @since 2.2
+ * @deprecated 2.7.0
+ *
+ * @param string $handle name.
+ * @return string the log file path.
+ */
+function wc_get_log_file_path( $handle ) {
```
Comment (claudiosanches) [inline includes/wc-deprecated-functions.php]: I'm not a big fan of deprecated a function in favor of a static method.
The use of this function is more easy than remember the class name and method.
@mikejolley thoughts?

Comment (sirreal) [inline includes/wc-deprecated-functions.php]: `WC_Log_Handler_File` is heavily dependent on this functionality, whereas `wc_get_log_file_path` outside of the context of the file log handler doesn't seem relevant.

That was the motivation for moving it to the class.

Comment (claudiosanches) [inline includes/wc-deprecated-functions.php]: Not saying to use inside the `WC_Log_Handler_File` this function or stop use the method, just saying that it's not required to deprecated this function and not seems good to deprecated a function forcing use a static method.

Comment (mikejolley) [inline includes/wc-deprecated-functions.php]: Hmm actually I prefer having functions available too yes. Plus it saves us affecting more devs with deprecation :p

Comment (sirreal) [inline includes/wc-deprecated-functions.php]: 👍  Done!

**Pull request merged**